cpp/tests/iterator/iterator_test.cu

```diff
@@ -0,0 +1,393 @@
+        RMM_TRY(RMM_ALLOC(&d_temp_storage, temp_storage_bytes, 0));
+
+        // Run reduction
+        cub::DeviceReduce::Reduce(d_temp_storage, temp_storage_bytes, d_in, dev_result.begin(), num_items,
+            cudf::DeviceSum{}, init);
+
+        evaluate(expected, dev_result, "cub test");
+    }
+
+    // iterator test case which uses thrust
+    template <typename InputIterator, typename T_output>
+    void iterator_test_thrust(T_output expected, InputIterator d_in, int num_items, bool is_device=true)
+    {
+        T init = T{0};
+
+        InputIterator d_in_last =  d_in + num_items;
+
+        EXPECT_EQ( thrust::distance(d_in, d_in_last), num_items);
+
+
+        T result;
+        if( is_device){
+            result = thrust::reduce(thrust::device, d_in, d_in_last, init, cudf::DeviceSum{});
+        }else{
+            for(auto it = d_in; it != d_in_last; ++it){
+                std::cout << "V: " << *it << std::endl;
+            }
+
+            std::cout << "thrust test (host start)" << std::endl;
+            result = thrust::reduce(thrust::host, d_in, d_in_last, init, cudf::DeviceSum{});
+            std::cout << "thrust test (host end)" << std::endl;
+        }
+
+        EXPECT_EQ(expected, result) << "thrust test";
+    }
+
+    void evaluate(T expected, thrust::device_vector<T> &dev_result, const char* msg=nullptr)
+    {
+        thrust::host_vector<T>  hos_result(dev_result);
+
+        EXPECT_EQ(expected, dev_result[0]) << msg ;
+        std::cout << "Done: expected <" << msg << "> = " << dev_result[0] << std::endl;
+    }
+};
+
+using TestingTypes = ::testing::Types<
+    int32_t
+>;
+
+TYPED_TEST_CASE(IteratorTest, TestingTypes);
+
+
+// tests for non-null iterator (pointer of device array)
+TYPED_TEST(IteratorTest, non_null_iterator)
+{
+    using T = int32_t;
+    std::vector<T> hos_array({0, 6, 0, -14, 13, 64, -13, -20, 45});
+    thrust::device_vector<T> dev_array(hos_array);
+
+    // calculate the expected value by CPU.
+    T expected_value = std::accumulate(hos_array.begin(), hos_array.end(), T{0});
+
+    // driven by iterator as a pointer of device array.
+    auto it_dev = dev_array.begin();
+    this->iterator_test_cub(expected_value, it_dev, dev_array.size());
+    this->iterator_test_thrust(expected_value, it_dev, dev_array.size());
+}
+
+
+/* tests for null input iterator (column with null bitmap)
+   Actually, we can use cub for reduction with nulls without creating custom kernel or multiple steps.
+   we may accelarate the reduction for a column using cub
+*/
+TYPED_TEST(IteratorTest, null_iterator)
+{
+    using T = int32_t;
+    T init = T{0};
+
+    std::vector<T> hos_array({0, 6, 0, -14, 13, 64, -13, -20, 45});
+    thrust::device_vector<T> dev_array(hos_array);
+
+    std::vector<bool> host_bools({1, 1, 0, 1, 1, 1, 0, 1, 1});
+    std::vector<gdf_valid_type> host_nulls;
+    gen_nullbitmap(host_nulls, host_bools);
+    thrust::device_vector<gdf_valid_type> dev_nulls(host_nulls);
```

Comment:
I think this would simpler if you used `column_wrapper`.

Reply:
Here's an example: https://github.com/rapidsai/cudf/blob/branch-0.8/cpp/tests/stream_compaction/stream_compaction_tests.cu

Reply:
sure.

Reply:
@jrhemstad @harrism 
Is it possible to get `column_wrapper::bitmask` to create `host_nulls`?
I also test reduction with `thrust::host `, `host_nulls` is required for testing host memory iterator.

Reply:
self resolved.
`column_wrapper::to_host()` would work.